internal/scheduler/upsert.go

```diff
@@ -0,0 +1,323 @@
+		)
+		if i != len(err.StaleWrites)-1 {
+			sb.WriteString(", ")
+		}
+	}
+	return sb.String()
+}
+
+// IdempotentUpsert is an optimized SQL call for bulk upserts.
+//
+// For efficiency, this function:
+// 1. Creates an empty temporary SQL table.
+// 2. Inserts all records into the temporary table using the postgres-specific COPY wire protocol.
+// 3. Upserts all records from the temporary table into the target table (as specified by tableName).
+//
+// The COPY protocol can be faster than repeated inserts for as little as 5 rows; see
+// https://www.postgresql.org/docs/current/populate.html
+// https://pkg.go.dev/github.com/jackc/pgx/v4#hdr-Copy_Protocol
+//
+// The records to write should be structs with fields marked with "db" tags.
+// Field names and values are extracted using the NamesValuesFromRecord function;
+// see its definition for details. The first field is used as the primary key in SQL.
+//
+// The temporary table is created with the provided schema, which should be of the form
+// (
+//
+//	id UUID PRIMARY KEY,
+//	width int NOT NULL,
+//	height int NOT NULL
+//
+// )
+// I.e., it should omit everything before and after the "(" and ")", respectively.
+//
+// This function relies on Pulsar message ids for idempotent writes;
+// if a Pulsar message id stored in postgres is more recent than one provided to this function,
+// the transaction is rolled back and a ErrStaleWrite error is returned.
+//
+// This function assumes all records are derived from a single (possibly partitioned) Pulsar topic
+// with name topicName. writeMessageIds maps partition indices to the id of the most recently
+// received Pulsar message for that partition.
+func IdempotentUpsert(ctx context.Context, db *pgxpool.Pool, topicName string, writeMessageIds map[int32]pulsar.MessageID, tableName string, schema string, records []interface{}) error {
+	if len(records) == 0 {
+		return nil
+	}
+	return db.BeginTxFunc(ctx, pgx.TxOptions{
+		IsoLevel:       pgx.ReadCommitted,
+		AccessMode:     pgx.ReadWrite,
+		DeferrableMode: pgx.Deferrable,
+	}, func(tx pgx.Tx) error {
+
+		err := IdempotencyCheck(ctx, tx, topicName, writeMessageIds)
+		if err != nil {
+			return err
+		}
+
+		err = CopyProtocolUpsert(ctx, tx, tableName, schema, records)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// Upsert is like [IdempotentUpsert], except the it does no idempotency check.
+func Upsert(ctx context.Context, db *pgxpool.Pool, tableName string, schema string, records []interface{}) error {
+	if len(records) == 0 {
+		return nil
+	}
+	return db.BeginTxFunc(ctx, pgx.TxOptions{
+		IsoLevel:       pgx.ReadCommitted,
+		AccessMode:     pgx.ReadWrite,
+		DeferrableMode: pgx.Deferrable,
+	}, func(tx pgx.Tx) error {
+		return CopyProtocolUpsert(ctx, tx, tableName, schema, records)
+	})
+}
+
+func IdempotencyCheck(ctx context.Context, tx DBTX, topicName string, writeMessageIds map[int32]pulsar.MessageID) error {
```

Comment:
What happens if this check fails?

Reply:
More specifically what happens if this check fails- but some of the messageIds are greater than the last storedid and thus need to be applied.